clippy_utils/src/ty.rs

```diff
@@ -1290,3 +1242,101 @@ pub fn normalize_with_regions<'tcx>(tcx: TyCtxt<'tcx>, param_env: ParamEnv<'tcx>
 pub fn is_manually_drop(ty: Ty<'_>) -> bool {
     ty.ty_adt_def().map_or(false, AdtDef::is_manually_drop)
 }
+
+#[derive(Clone, Copy)]
+pub enum Sizedness {
+    /// The type is uninhabited. (e.g. `!`)
+    Uninhabited,
+    /// The type is zero-sized.
+    Zero,
+    /// The type has some other size or an unknown size.
+    Other,
+}
+impl Sizedness {
+    pub fn is_zero(self) -> bool {
+        matches!(self, Self::Zero)
+    }
+
+    pub fn is_uninhabited(self) -> bool {
+        matches!(self, Self::Uninhabited)
+    }
+}
+
+/// Calculates the sizedness of a type.
+pub fn sizedness_of<'tcx>(tcx: TyCtxt<'tcx>, param_env: ParamEnv<'tcx>, ty: Ty<'tcx>) -> Sizedness {
+    fn for_list<'tcx>(
+        tcx: TyCtxt<'tcx>,
+        param_env: ParamEnv<'tcx>,
+        tys: impl IntoIterator<Item = Ty<'tcx>>,
+    ) -> Sizedness {
+        let mut res = Sizedness::Zero;
+        for ty in tys {
+            match sizedness_of(tcx, param_env, ty) {
+                Sizedness::Uninhabited => return Sizedness::Uninhabited,
+                Sizedness::Other => res = Sizedness::Other,
+                Sizedness::Zero => {},
+            }
+        }
+        res
+    }
+
+    match *ty.kind() {
+        ty::FnDef(..) => Sizedness::Zero,
+        ty::Tuple(tys) => for_list(tcx, param_env, tys),
+        ty::Array(_, len) if len.eval_target_usize(tcx, param_env) == 0 => Sizedness::Zero,
+        ty::Array(ty, _) => sizedness_of(tcx, param_env, ty),
+        ty::Adt(adt, args) => {
+            let mut iter = adt
+                .variants()
+                .iter()
+                .map(|v| for_list(tcx, param_env, v.fields.iter().map(|f| f.ty(tcx, args))))
+                .filter(|x| !x.is_uninhabited());
+            match iter.next() {
+                None => Sizedness::Uninhabited,
+                Some(Sizedness::Other) => Sizedness::Other,
+                Some(_) => {
+                    if iter.next().is_some() {
+                        Sizedness::Other
+                    } else {
+                        Sizedness::Zero
+                    }
+                },
+            }
+        },
+        ty::Closure(_, args) => for_list(tcx, param_env, args.as_closure().upvar_tys().iter()),
+        ty::Coroutine(_, args) => for_list(tcx, param_env, args.as_coroutine().upvar_tys().iter()),
+        ty::CoroutineClosure(_, args) => for_list(tcx, param_env, args.as_coroutine_closure().upvar_tys().iter()),
+        ty::CoroutineWitness(_, args) => for_list(tcx, param_env, args.iter().filter_map(GenericArg::as_type)),
```

Comment:
I don't think coroutines are ZSTs if they don't have upvars. At least futures generated by async blocks are always at least of size 1 to tell that they've finished https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=743d622d9887740234d325ae94dbf5f4

Then again, I have no idea if `zero_sized_map_values` realistically could ever run into this?

Reply:
That is a good point. Fixed.

The check for an uninhabited coroutine is still conservative, but that should be fine.